certbot-nginx/certbot_nginx/_internal/configurator.py

```diff
@@ -723,41 +724,62 @@ def _make_server_ssl(self, vhost: obj.VirtualHost) -> None:
 
         """
         https_port = self.config.https_port
-        ipv6info = self.ipv6_info(str(https_port))
-        ipv6_block = ['']
-        ipv4_block = ['']
+        http_port = self.config.http01_port
+
+        # no addresses should have ssl turned on here
+        assert not vhost.ssl
+
+        sslified_addrs: List[obj.Addr] = [obj.Addr(addr.get_addr(), str(https_port), True, False,
+                                               addr.ipv6, False)
+                                         for addr in vhost.addrs
+                                         if addr.get_port() == str(http_port)]
 
         # If the vhost was implicitly listening on the default Nginx port,
         # have it continue to do so.
         if not vhost.addrs:
             listen_block = [['\n    ', 'listen', ' ', self.DEFAULT_LISTEN_PORT]]
             self.parser.add_server_directives(vhost, listen_block)
 
-        if vhost.ipv6_enabled():
-            ipv6_block = ['\n    ',
-                          'listen',
-                          ' ',
-                          '[::]:{0}'.format(https_port),
-                          ' ',
-                          'ssl']
-            if not ipv6info[1]:
-                # ipv6only=on is absent in global config
-                ipv6_block.append(' ')
-                ipv6_block.append('ipv6only=on')
-
-        if vhost.ipv4_enabled():
-            ipv4_block = ['\n    ',
-                          'listen',
-                          ' ',
-                          '{0}'.format(https_port),
-                          ' ',
-                          'ssl']
+        if not sslified_addrs:
+            # there are no existing addressing listening on 80
+            if vhost.ipv6_enabled():
+                sslified_addrs += [obj.Addr('[::]', str(https_port), True, False, True, False)]
+            if vhost.ipv4_enabled():
+                sslified_addrs += [obj.Addr('', str(https_port), True, False, False, False)]
+
+        addr_blocks: List[List[str]] = []
+        ipv6only_set_here: Set[Tuple[str, str]] = set()
+        for addr in sslified_addrs:
+            host = addr.get_addr()
+            port = addr.get_port()
+            if addr.ipv6:
+                addr_block = ['\n    ',
```

Comment:
Also debated setting `ipv6only` in `addr`, getting the string representation, splitting the result on spaced, and adding spaces back in, but that seemed messier.

Reply:
i'm fine w/ either, though this way is at least a bit more predictable in what we get as output. but why do we even have to include the whitespace here??

**edit** i'm seeing it's because of the parser api. awful

Reply:
Yeah. The whole unspaced list shenanigans. Which is mostly awful because we need to keep that data around, less an implementation choice.